bindings/matrix-sdk-ffi/src/sliding_sync.rs

```diff
@@ -0,0 +1,582 @@
+                .unwrap_or_default(),
+        }
+    }
+}
+
+pub struct SlidingSyncRoom {
+    inner: matrix_sdk::SlidingSyncRoom,
+    client: Client,
+}
+
+impl SlidingSyncRoom {
+    pub fn name(&self) -> Option<String> {
+        self.inner.name().map(ToOwned::to_owned)
+    }
+
+    pub fn room_id(&self) -> String {
+        self.inner.room_id().to_string()
+    }
+
+    pub fn is_dm(&self) -> Option<bool> {
+        self.inner.is_dm
+    }
+
+    pub fn is_initial(&self) -> Option<bool> {
+        self.inner.initial
+    }
+    pub fn is_loading_more(&self) -> bool {
+        self.inner.is_loading_more()
+    }
+
+    pub fn has_unread_notifications(&self) -> bool {
+        !self.inner.unread_notifications.is_empty()
+    }
+
+    pub fn unread_notifications(&self) -> Arc<UnreadNotificationsCount> {
+        Arc::new(self.inner.unread_notifications.clone().into())
+    }
+
+    #[allow(clippy::significant_drop_in_scrutinee)]
+    pub fn latest_room_message(&self) -> Option<Arc<crate::messages::AnyMessage>> {
+        let messages = self.inner.timeline();
+        // room is having the latest events at the end,
+        let lock = messages.lock_ref();
+        for m in lock.iter().rev() {
+            if let Some(e) = crate::messages::sync_event_to_message(m.clone().into()) {
+                return Some(e);
+            }
+        }
+        None
+    }
+
+    pub fn full_room(&self) -> Option<Arc<super::Room>> {
+        self.client.get_room(self.inner.room_id()).map(|room| Arc::new(super::Room::new(room)))
+    }
+}
+
+pub struct UpdateSummary {
+    /// The views (according to their name), which have seen an update
+    pub views: Vec<String>,
+    pub rooms: Vec<String>,
+}
+
+pub struct RequiredState {
+    pub key: String,
+    pub value: String,
+}
+
+pub struct RoomSubscription {
+    pub required_state: Option<Vec<RequiredState>>,
+    pub timeline_limit: Option<u32>,
+}
+
+impl TryInto<RumaRoomSubscription> for RoomSubscription {
+    type Error = anyhow::Error;
+    fn try_into(self) -> anyhow::Result<RumaRoomSubscription> {
+        Ok(assign!(RumaRoomSubscription::default(), {
+            required_state: self.required_state.map(|r|
+                r.into_iter().map(|s| (s.key.into(), s.value)).collect()
+            ).unwrap_or_default(),
+            timeline_limit: self.timeline_limit.map(|u| u.into())
+        }))
+    }
+}
+
+impl From<matrix_sdk::UpdateSummary> for UpdateSummary {
+    fn from(other: matrix_sdk::UpdateSummary) -> UpdateSummary {
+        UpdateSummary {
+            views: other.views,
+            rooms: other.rooms.into_iter().map(|r| r.as_str().to_owned()).collect(),
+        }
+    }
+}
+
+pub enum SlidingSyncViewRoomsListDiff {
+    Replace { values: Vec<RoomListEntry> },
+    InsertAt { index: u32, value: RoomListEntry },
+    UpdateAt { index: u32, value: RoomListEntry },
+    RemoveAt { index: u32 },
+    Move { old_index: u32, new_index: u32 },
```

Comment:
I guess `old_index` means the `from_index`/`source_index` and `new_index` means `to_index`/`destination_index`?

I don't know which vocabulary I prefer the most, but “old”/“new” doesn't feel appropriate. I can live with it though, it's private data, then it's fairly easy to change. I let you decide!

PS: I reckon this naming comes from `VecDiff`?

Reply:
Yeah the naming comes from `VecDiff`.

Reply:
I'll leave it as is as this comes from the [`VecDiff`](https://docs.rs/futures-signals/latest/futures_signals/signal_vec/enum.VecDiff.html#variant.Move) dependency directly.